plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi_test.go

```diff
@@ -0,0 +1,239 @@
+		return nil
+	case 2:
+		notification := mockGNMINotification()
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		return nil
+	case 3:
+		notification := mockGNMINotification()
+		notification.Prefix.Elem[0].Key["foo"] = "bar2"
+		notification.Update[0].Path.Elem[1].Key["name"] = "str2"
+		notification.Update[0].Val = &gnmi.TypedValue{Value: &gnmi.TypedValue_BoolVal{BoolVal: false}}
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		return nil
+	default:
+		return fmt.Errorf("test not implemented ;)")
+	}
+}
+
+func TestGNMIError(t *testing.T) {
+	listener, _ := net.Listen("tcp", "127.0.0.1:57003")
+	server := grpc.NewServer()
+	gnmi.RegisterGNMIServer(server, &mockGNMIServer{t: t, scenario: 0, server: server})
+
+	c := &CiscoTelemetryGNMI{Addresses: []string{"127.0.0.1:57003"},
+		Username: "theuser", Password: "thepassword", Encoding: "proto",
+		Redial: internal.Duration{Duration: 1 * time.Second}}
+
+	acc := &testutil.Accumulator{}
+	assert.Nil(t, c.Start(acc))
+	server.Serve(listener)
+	c.Stop()
+
+	assert.Contains(t, acc.Errors, errors.New("aborted GNMI subscription: rpc error: code = Unknown desc = testerror"))
+}
+
+func mockGNMINotification() *gnmi.Notification {
+	return &gnmi.Notification{
+		Timestamp: 1543236572000000000,
+		Prefix: &gnmi.Path{
+			Origin: "type",
+			Elem: []*gnmi.PathElem{
+				{
+					Name: "model",
+					Key:  map[string]string{"foo": "bar"},
+				},
+			},
+			Target: "subscription",
+		},
+		Update: []*gnmi.Update{
+			{
+				Path: &gnmi.Path{
+					Elem: []*gnmi.PathElem{
+						{Name: "some"},
+						{
+							Name: "path",
+							Key:  map[string]string{"name": "str", "uint64": "1234"}},
+					},
+				},
+				Val: &gnmi.TypedValue{Value: &gnmi.TypedValue_IntVal{IntVal: 5678}},
+			},
+			{
+				Path: &gnmi.Path{
+					Elem: []*gnmi.PathElem{
+						{Name: "other"},
+						{Name: "path"},
+					},
+				},
+				Val: &gnmi.TypedValue{Value: &gnmi.TypedValue_StringVal{StringVal: "foobar"}},
+			},
+			{
+				Path: &gnmi.Path{
+					Elem: []*gnmi.PathElem{
+						{Name: "other"},
+						{Name: "this"},
+					},
+				},
+				Val: &gnmi.TypedValue{Value: &gnmi.TypedValue_StringVal{StringVal: "that"}},
+			},
+		},
+	}
+}
+
+func TestGNMIMultiple(t *testing.T) {
+	listener, _ := net.Listen("tcp", "127.0.0.1:57004")
+	server := grpc.NewServer()
+	gnmi.RegisterGNMIServer(server, &mockGNMIServer{t: t, scenario: 1, server: server})
+
+	c := &CiscoTelemetryGNMI{Addresses: []string{"127.0.0.1:57004"},
+		Username: "theuser", Password: "thepassword", Encoding: "proto",
+		Redial:        internal.Duration{Duration: 1 * time.Second},
+		Subscriptions: []Subscription{{Name: "alias", Origin: "type", Path: "/model", SubscriptionMode: "sample"}},
+	}
+
+	acc := &testutil.Accumulator{}
+	assert.Nil(t, c.Start(acc))
+
+	server.Serve(listener)
+	c.Stop()
```

Comment:
I think this will be racy, in case the data is not received before Stop is called.  You can use the acc.Wait(n) function to wait for a known number of metrics.

Reply:
Done, thanks!